server/controllers/duels.js

```diff
@@ -0,0 +1,600 @@
+ *
+ * @apiSuccess (200) {object} duel The updated duel
+ *
+ * @apiParamExample  {type} Request-Example:
+ * {
+ *     "answers" : [1,3,0,0,2]
+ * }
+ *
+ * @apiUse ErrorServer
+ * @apiError (404) NotFound Duel does not exist
+ * @apiError (400) InvalidRound The round is invalid
+ * @apiError (400) FinishedDuel The duel is finished
+ * @apiError (400) AlreadyPlayed The user has already played this round
+ * @apiError (400) InvalidAnswers The user answers are invalid
+ *
+ *
+ */
+function play(req, res) {
+  const id = Number(req.params.id);
+  const round = Number(req.params.round);
+  const username = req.body.auth_user;
+  const answers = req.body.answers || [];
+
+  if (!id) {
+    return res.status(400).json({ message: "Invalid or missing duel id" });
+  }
+
+  getDuel(id, username).then((duel) => {
+    if (!duel) {
+      return res.status(404).json({ message: "Duel not found" });
+    }
+
+    if (!duel.inProgress) {
+      return res.status(400).json({ message: "This duel is finished" });
+    }
+
+    if (duel.currentRound !== round) {
+      return res.status(400).json({ message: "Invalid duel round" });
+    }
+    if (duel.rounds[round - 1][0].userAnswer !== undefined) {
+      return res.status(400).json({ message: "You can only play a round once" });
+    }
+
+    if (answers.length !== duel.rounds[round - 1].length) {
+      return res.status(400).json({ message: "Incorrect number of answers" });
+    }
+
+    const sendLocalError500 = (error) => sendError500(res, error);
+
+    insertResultInDatabase(id, username, answers)
+      .then((newDuel) =>
+        updateDuelState(newDuel, username)
+          .then((duel) => res.status(200).json(duel))
+          .catch(sendLocalError500)
+      )
+      .catch(sendLocalError500);
+  });
+}
+
+let mockedDuelsRounds;
+/**
+ * Function to control the generation of random rounds by passing a predefined set of rounds.
+ * Use only for testing purposes.
+ * @param {object} fakeDuelsRounds The predefined set of rounds
+ */
+export function _initMockedDuelRounds(fakeDuelsRounds) {
+  if (process.env.NODE_ENV !== "test") {
+    throw "function reserved for tests";
+  }
+  mockedDuelsRounds = fakeDuelsRounds;
+}
+
+export default { create, fetch, fetchAll, play };
+
+// ***** INTERNAL FUNCTIONS *****
+
+/**
+ * while waiting for a global error handler
+ */
+function sendError500(res, error) {
+  if (NotEnoughDataError.isInstance(error)) {
+    return res.status(422).json({ message: error.message, code: error.code });
+  }
+  console.error(error);
+  return res.status(500).json({ message: "Server side error" });
+}
+
+/**
+ * Check if all users in a list exist
+ * @param  {...string} users The list of users
+ * @returns {Promise<boolean>}
+ */
+function doUsersExist(...users) {
+  return new Promise((resolve, reject) => {
+    const sql = ` SELECT us_login \
+                  FROM user \
+                  WHERE us_login IN (${Array(users.length).fill("?").join(",")});`;
```

Comment:
You could make a `COUNT()` query instead of selecting all the logins just to count the number of rows :-)